rust/src/types/stack.rs

```diff
@@ -1,19 +1,112 @@
+#[cfg(feature = "stack-array")]
+use std::{cmp::min, mem::MaybeUninit};
+
 use crate::types::{u256, FailStatus};
 
+#[cfg(feature = "stack-array")]
+#[derive(Debug)]
+struct StackRaw {
+    data: [MaybeUninit<u256>; 1024],
+    len: usize,
+}
+
+#[cfg(feature = "stack-array")]
+impl StackRaw {
+    pub fn new(inner: &[u256]) -> Self {
+        let len = min(inner.len(), 1024);
+        let mut s = Self {
+            data: [const { MaybeUninit::uninit() }; 1024],
+            len,
+        };
+        // SAFETY:
+        // &[T] and &[MaybeUninit<T>] have the same layout
+        // With nightly rust this could be replaced by MaybeUninit::copy_from_slice
+        s.data[..len].copy_from_slice(unsafe {
+            std::mem::transmute::<&[u256], &[std::mem::MaybeUninit<u256>]>(inner)
+        });
+        s
+    }
+
+    pub fn as_slice(&self) -> &[u256] {
+        // SAFETY:
+        // &[T] and &[MaybeUninit<T>] have the same layout and self.data is initialized up to
+        // self.len
+        unsafe {
+            std::mem::transmute::<&[std::mem::MaybeUninit<u256>], &[u256]>(&self.data[..self.len])
+        }
+    }
+
+    pub fn as_mut_slice(&mut self) -> &mut [u256] {
+        // SAFETY:
+        // &[T] and &[MaybeUninit<T>] have the same layout and self.data is initialized up to
+        // self.len
+        unsafe {
+            std::mem::transmute::<&mut [std::mem::MaybeUninit<u256>], &mut [u256]>(
+                &mut self.data[..self.len],
+            )
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        self.len
+    }
+
+    pub unsafe fn push(&mut self, value: u256) {
+        *self.data.get_unchecked_mut(self.len) = MaybeUninit::new(value);
+        self.len += 1;
+    }
+
+    pub fn pop_array<const N: usize>(&mut self) -> Option<[u256; N]> {
+        if self.len < N {
```

Comment:
dead code?  check overflow should prevent this.

Reply:
Yes. I think pop_array should be unsafe and assume that the check was already done and then return [u256; N] directly.